[Clang] Add "extend lifetime" flags and release note

SLTozer opened the pull request:
Following the commit that added the fake use intrinsic to LLVM, this patch adds a pair of flags for the clang frontend that emit fake use intrinsics, for the purpose of extending the lifetime of variables (either all source variables, or just the `this` pointer). This patch does not implement the fake use intrinsic emission of the flags themselves, it simply adds the flags, the corresponding release note, and the attachment of the `has_fake_uses` attribute to affected functions; the remaining functionality appears in the next patch.

All code originally written by @wolfy1961, while I'll be handling the review.

SLTozer committed:
[Clang] Add "extend lifetime" flags and release note

Following the commit that added the fake use intrinsic to LLVM, this patch
adds a pair of flags for the clang frontend that emit fake use intrinsics,
for the purpose of extending the lifetime of variables (either all source
variables, or just the `this` pointer). This patch does not implement the
fake use intrinsic emission of the flags themselves, it simply adds the flags,
the corresponding release note, and the attachment of the `has_fake_uses`
attribute to affected functions; the remaining functionality appears in the
next patch.

SLTozer commented:
Ping - if anyone isn't able to review themselves but has an idea of someone who might, adding/pinging them would also be appreciated.

dwblaikie commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -211,6 +211,16 @@ New Compiler Flags
+  but applies only to the ``this`` variable in C++ class member functions. Note
+  that this flag modifies the optimizations that Clang performs, which will
+  result in reduced performance in generated code; however, this feature will
+  not extend the lifetime of some variables in cases where doing so would have
```
huh - what cases is this alluding to?

SLTozer replied:
The main case off the top of my head is that we drop fake uses for inlined functions; preserving them results in a fairly significant drop in performance, enough to hurt the flag's viability for game debugging. Though now that this is coming upstream, it's possible we could try toggling that behaviour (and any similar cases) depending on opt level; in the [earlier RFC](https://discourse.llvm.org/t/rfc-redefine-og-o1-and-add-a-new-level-of-og/72850) I proposed having an extra optimization mode that aims for high performance (around 5% 
slower than O2 for some representative benchmarks) while having good debug info - an appropriate solution may be that we'd drop inlined fake uses at that opt level, while keeping them at Og, for example.

pogo59 reviewed (commented):
It seems a little odd to have only positive forms of these flags; usually toggles would have both positive and negative forms. Maybe @MaskRay has an opinion?

The release note doesn't say: Does `-fextend-lifetimes` imply `-fextend-this-pointer`? They're implemented as independent toggles but the effect isn't really independent IIRC. I wonder (years after it was originally implemented downstream, I know) whether we'd be better off with `-fextend-lifetimes[={all,this,none}]`. (But maybe wait for a second opinion.)

dwblaikie commented on clang/lib/Frontend/CompilerInvocation.cpp:
```diff
@@ -2217,6 +2217,11 @@ bool CompilerInvocation::ParseCodeGenArgs(CodeGenOptions &Opts, ArgList &Args,
                       Opts.SanitizeTrap);
 
+  Opts.ExtendThisPtr =
+      Opts.OptimizationLevel > 0 && Args.hasArg(OPT_fextend_this_ptr);
```
Is it worth maybe making this inconditional, even at -O0, in case someone's put an optimize attribute on a function, etc, and wants to override it?

SLTozer replied:
I'm currently looking into this to see if there would be any notable negative consequences, but for now I'm at least happy to move this to the [next patch along](https://github.com/llvm/llvm-project/pull/110102): in that patch, we emit fake uses for functions when `OptimizationLevel != 0` and _or_ we've passed `-disable-O0-optnone` - that at least gives the ability to pass `-O0 -fextend-lifetimes` together if desired.

dwblaikie commented on clang/test/CodeGen/extend-lifetimes-hasfakeuses.c:
```diff
@@ -0,0 +1,12 @@
+// when optimizations are disabled.
+
+// CHECK-ALL:    define {{.*}}void @foo
+// CHECK-O2:     attributes #0 = {{{.*}}has_fake_uses
```
Interesting model - I haven't reviewed the other patches in this direction, but I'm a bit surprised this is the only IR effect from the frontend.

I take it this attribute is added, which then causes some middle end pass to add the fake uses themselves?

That seems a bit quirky in terms of creating an attribute to communicate that - I guess other directions have already been considered/found unsuitable? (a couple of others that come to mind include: having the frontend create the fake uses (limits reusability for other front ends) or having the frontend choose to insert the fake use pass or not (and then the pass is unconditional/doesn't look for an attribute) - I /think/ a combination of these might be how the sanitizers work? (frontend chooses to include certain passes but might also have to do some different IRGen too))

SLTozer replied:
This isn't the only effect from the front end; this flag will actually generate llvm.fake.use intrinsics, but only for source variables, which this test does not have. But as you've noted, this test should probably include variables to demonstrate that we also don't emit fake uses for those at O0!

SLTozer commented:
> The release note doesn't say: Does `-fextend-lifetimes` imply `-fextend-this-pointer`? They're implemented as independent toggles but the effect isn't really independent IIRC. I wonder (years after it was originally implemented downstream, I know) whether we'd be better off with `-fextend-lifetimes[={all,this,none}]`. (But maybe wait for a second opinion.)

Correct, this is _sort of_ implied by `-fextend-lifetimes` extending the life of all variables and `-fextend-this-pointer` extending just the `this` pointer, but it wasn't totally clear to me the first time I looked at the flag either. Making it into a flag that takes options is an interesting idea and one I broadly approve of - especially as I have some plans to add additional modes (e.g. `-fextend-lifetimes=params`) further down the line.

SLTozer committed:
Make options into BoolF types, remove added attribute

SLTozer commented:
This patch has been changed as a result of some other patches that have removed the need for the attribute that this patch once added. Since this patch does not emit fake uses (the actual emission of fake uses is added in the [next patch](https://github.com/llvm/llvm-project/pull/110102)), that leaves this patch in the awkward position of adding flags and a release note but not actually doing anything. I'd prefer to leave this patch separate though, because the next patch is already tricky to review - adding more to that patch would probably slow things down a bit.

SLTozer committed:
Make extend lifetimes an = option

jmorse commented:
The additions here look fine; however I think there's generally precedent that anything we add needs to have /some/ kind of test, or be covered by something already existing, otherwise we're vulnerable to:
 * Patch lands,
 * Someone refactors clang switch handling,
 * Other patches land but the flag doesn't do anything / is rejected,
 * Confusion reigns

As far as I understand it, these are driver options that will be passed through to cc1? In that case we can at least test the passthrough, i.e. that `-fextend-lifetimes` appears in the output of `-###`.

SLTozer commented:
> As far as I understand it, these are driver options that will be passed through to cc1? In that case we can at least test the passthrough, i.e. that `-fextend-lifetimes` appears in the output of `-###`.

I think my preferred approach would be to merge both patches simultaneously, or else roll this patch into the next one - we have some tests that use `-###` to test some meaningful logic, but this would essentially be testing that TableGen hasn't bugged out (for this specific flag).

zygoloid reviewed (commented):
I think this flag name is going to be extremely confusing. "[lifetime](https://en.cppreference.com/w/cpp/language/lifetime)" is a notion [defined formally by the C++ language standard](https://eel.is/c++draft/basic.life), and "lifetime extension" is again a process [defined in C++](https://eel.is/c++draft/class.temporary#6) with a specific meaning. Those meanings have nothing to do with what these flags control.

Please consider alternative names for this flag that make it clear that it's not talking about changing language semantics, just how code is lowered. Maybe `-fextend-live-ranges` or `-fkeep-alive` or something like that? (Maybe even something that mentions debugging as the intended use case, as opposed to, say, supporting stack-walking GC?)

zygoloid commented:
Looks like the tests in #110102 are named `extend-liveness-*.cpp`. Perhaps `-fextend-liveness` would be a good name for this flag?

SLTozer committed:
Rename to -fextend-variable-liveness

SLTozer committed:
Add visibility to alias

SLTozer commented:
I've tried renaming the option to `-fextend-variable-liveness` to make it more explicit and avoid confusion with C++ lifetimes, and also added an = option as suggested by @pogo59.

SLTozer committed:
Add Driver test for flag

SLTozer committed:
Improve help text for new flags

SLTozer commented:
I've added a test for this patch that tests the driver - since we will now be relying on the driver to decide whether to pass these flags to the frontend, rather than trivially always doing so (so that [later on](https://github.com/llvm/llvm-project/pull/118026) we can use the driver to enable this at `-Og`), having a test seems useful. Also updated the help text to give a little more detail on usage and explain some key limitations.

SLTozer commented:
Post-holiday ping - I believe all the comments have been addressed now, does the new form look acceptable?

zygoloid commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -412,6 +412,20 @@ New Compiler Flags
+  through optimizations, meaning that variables will typically be visible in a
+  debugger more often. The flag has two levels: ``-fextend-variable-liveness``,
+  or ``-fextend-variable-liveness=all``, extendes the liveness of all user
+  variables and the ``this`` pointer. Alternatively ``-fextend-this-ptr``, or
```
Looks like the flag name supported is `-fextend-this-ptr-liveness`, not `-fextend-this-ptr`.

I also wonder whether the `-fextend-this-ptr-liveness` flag carries its weight -- we've historically removed these special-case flags in favor of the more general things, as we did with `-fsanitize=` for example. I understand that this is being upstreamed from some existing deployment; does the alias help there? Having it around for a shorter time period with an intent to eventually remove it might be reasonable.

SLTozer replied:
Sounds reasonable to me - the `this` variant is quite useful for more performance-sensitive cases, but it doesn't need to exist under a separate flag name; changed it.

zygoloid reviewed (commented):
New flag name LGTM.

SLTozer committed:
Remove -fextend-this-ptr alias

SLTozer committed:
Shorten help text

SLTozer commented:
Merged in [71ab44a8](https://github.com/llvm/llvm-project/commit/71ab44a8193c56e4ef9aede4d9bac8e14760c6c6), via command line so as to provide proper commit author attribution to @wolfy1961 (as this doesn't seem to be possible via the web interface).
